# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/glyph_rasterizer/mod.rs
# Commit: 793e9dfa4b9f
# Full Hash: 793e9dfa4b9fa40137cdf2a5d34c126d6cc9840d
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2020-02-29 04:04:56
# Regressor Bug: 1616995
# File Overlap Count: 1
# Description:
#   Bug 1616995 - patch 2 - Support vertical skew for upright-vertical fonts in webrender. r=lsalzman
#   
#   Depends on D63893
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D64746
# ==============================================================================

diff -r 88d3197f16c2 -r 793e9dfa4b9f gfx/wr/webrender/src/glyph_rasterizer/mod.rs
--- a/gfx/wr/webrender/src/glyph_rasterizer/mod.rs	Fri Feb 28 14:55:39 2020 +0000
+++ b/gfx/wr/webrender/src/glyph_rasterizer/mod.rs	Fri Feb 28 15:36:19 2020 +0000
@@ -338,14 +338,27 @@
         self.pre_scale(x_scale.recip() as f32, y_scale.recip() as f32)
     }
 
-    pub fn synthesize_italics(&self, angle: SyntheticItalics) -> Self {
+    pub fn synthesize_italics(&self, angle: SyntheticItalics, size: f64, vertical: bool) -> (Self, (f64, f64)) {
         let skew_factor = angle.to_skew();
-        FontTransform::new(
-            self.scale_x,
-            self.skew_x - self.scale_x * skew_factor,
-            self.skew_y,
-            self.scale_y - self.skew_y * skew_factor,
-        )
+        if vertical {
+          // origin delta to be applied so that we effectively skew around
+          // the middle rather than edge of the glyph
+          let (tx, ty) = (0.0, -size * 0.5 * skew_factor as f64);
+          (FontTransform::new(
+              self.scale_x + self.skew_x * skew_factor,
+              self.skew_x,
+              self.skew_y + self.scale_y * skew_factor,
+              self.scale_y,
+          ), (self.scale_x as f64 * tx + self.skew_x as f64 * ty,
+              self.skew_y as f64 * tx + self.scale_y as f64 * ty))
+        } else {
+          (FontTransform::new(
+              self.scale_x,
+              self.skew_x - self.scale_x * skew_factor,
+              self.skew_y,
+              self.scale_y - self.skew_y * skew_factor,
+          ), (0.0, 0.0))
+        }
     }
 
     pub fn swap_xy(&self) -> Self {
@@ -552,6 +565,10 @@
             0
         }
     }
+
+    pub fn synthesize_italics(&self, transform: FontTransform, size: f64) -> (FontTransform, (f64, f64)) {
+        transform.synthesize_italics(self.synthetic_italics, size, self.flags.contains(FontInstanceFlags::VERTICAL))
+    }
 }
 
 #[repr(u32)]